src/pf_types.h

```diff
@@ -2421,6 +2436,253 @@ typedef struct pf_log_book
+    * @param crep             In:    The IOCR index.
+    * @return  0  if the CPM instance could be activated.
+    *          -1 if an error occurred.
+    */
+   int (*activate_req) (pnet_t * net, pf_ar_t * p_ar, uint32_t crep);
+
+   /**
+    * Close a CPM instance.
+    *
+    * This function terminates the specified CPM instance.
+    * De-registers a frame handler for incoming Ethernet frames.
+    * The buffer mutex is destroyed on the last call.
+    *
+    * @param net              InOut: The p-net stack instance
+    * @param p_ar             InOut: The AR instance.
+    * @param crep             In:    The IOCR index.
+    * @return  0  if the CPM instance was closed.
+    *          -1 if an error occurred.
+    */
+   int (*close_req) (pnet_t * net, pf_ar_t * p_ar, uint32_t crep);
+
+   /**
+    * Retrieve the specified sub-slot IOCS sent from the controller.
+    * User must supply a buffer large enough to hold the received IOCS.
+    * Maximum buffer size is 256 bytes.
+    * @param net           InOut: The p-net stack instance
+    * @param iocr          InOut:    The IOCR instance.
+    * @param p_iodata      In:    Iodata object information
+    * @param p_iocs        Out:   Copy of the received IOCS.
+    * @param iocs_len      In:    Size of IOCS.
+    * @return 0 on success, -1 on error
+    */
+   int (*get_iocs) (
+      pnet_t * net,
+      pf_iocr_t * iocr,
+      const pf_iodata_object_t * p_iodata,
+      uint8_t * p_iocs,
+      uint8_t iocs_len);
+
+   /**
+    * Retrieve the specified sub-slot data and IOPS received from the
+    * controller.
+    *
+    * User must supply a buffer large enough to hold the received
+    * data. Maximum buffer size is PF_FRAME_BUFFER_SIZE (1500) bytes.
+    *
+    * User must also supply a buffer large enough to hold the received IOPS.
+    * Maximum buffer size is 256 bytes.
+    *
+    * @param net           InOut: The p-net stack instance
+    * @param iocr          InOut: The IOCR instance.
+    * @param p_iodata      In:    Iodata object information
+    * @param p_new_flag    Out:   true means new valid data (and IOPS) frame
+    *                             available since last call.
+    * @param p_data        Out:   Copy of the received data.
+    * @param data_len      In:    Buffer size.
+    * @param p_iops        Out:   The received IOPS.
+    * @param iops_len      In:    Size of buffer at IOPS.
+    * @return  0  if the data and IOPS could be retrieved.
+    *          -1 if an error occurred.
+    */
+   int (*get_data_and_iops) (
+      pnet_t * net,
+      pf_iocr_t * iocr,
+      const pf_iodata_object_t * p_iodata,
+      bool * p_new_flag,
+      uint8_t * p_data,
+      uint16_t data_len,
+      uint8_t * p_iops,
+      uint8_t iops_len);
+
+   /**
+    * Get the data status of the CPM connection.
+    * @param p_cpm            In:   The CPM instance.
+    * @param p_data_status    Out:  The CPM data status.
```

Comment:
The CPM data status.
-> 
The CPM data status. See pnet_data_status_bits_t for details.

Reply:
OK